pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,815 @@
+func TestDispatch(t *testing.T) {
+	tests := []struct {
+		name       string
+		fields     Controller
+		message    *model.Message
+		wantErr    bool
+		isResponse bool
+	}{
+		{"dispatch with valid input", Controller{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.NewMessage("").BuildRouter(ModuleNameEdgeHub, core.TwinGroup, "", ""),
+			false, false},
+
+		{"Error Case in dispatch", Controller{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.EdgedGroup, "", ""),
+			true, true},
+
+		{"Response Case in dispatch", Controller{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.TwinGroup, "", ""),
+			false, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.isResponse && tt.wantErr != true {
+				keepChannel := ehc.addKeepChannel("test")
+				go ehc.dispatch(*tt.message)
+				got := <-keepChannel
+				if got != *tt.message {
+					t.Errorf("SendToKeep() error want: %v got: %v", *tt.message, got)
+				}
+			} else {
+				if err := ehc.dispatch(*tt.message); (err != nil) != tt.wantErr {
+					t.Errorf("TestController_dispatch() error = %v, wantErr %v", err, tt.wantErr)
+				}
+			}
+		})
+	}
+}
+
+//TestRouteToEdge() is used to test whether the message received from websocket is dispatched to the required modules
+func TestRouteToEdge(t *testing.T) {
+	initMocks(t)
+	tests := []struct {
+		name       string
+		fields     Controller
+		receiveErr bool
+	}{
+		{"Route to edge with proper input", Controller{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			chClient:   mockAdapter,
+			syncKeeper: make(map[string]chan model.Message),
+			stopChan:   make(chan struct{}),
+		}, false},
+
+		{"Receive Error in route to edge", Controller{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			chClient:   mockAdapter,
+			syncKeeper: make(map[string]chan model.Message),
+			stopChan:   make(chan struct{}),
+		}, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.receiveErr {
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage(""), errors.New("Connection Refused")).AnyTimes()
```

Comment:
can we parameterize the return and times of mock statements to remove if else here ?